test/parens-extra.ts

```diff
@@ -55,4 +54,20 @@ describe("babylon parens", function () {
 
     assert.strictEqual(printer.print(ast).code, "(1).foo");
   });
+
+  it("prints top level parens for an expression ast", function () {
+    check("(() => {})()");
+    check("(function () {} ())");
+  });
+
+  describe("reprinter", function () {
+    it("preserves necessary parens", function () {
+      const ast = parse("() => ({ prop: true })");
+      const expr = ast.program.body[0].expression;
+
+      expr.body.properties = [];
+
+      assert.strictEqual(printer.print(ast).code, "() => ({})");
```

Comment:
I haven't yet been able to verify that this test fails if the behavior regresses.